drivers/flash/flash_stm32_qspi.c

```diff
@@ -0,0 +1,927 @@
+	struct dma_config dma_cfg = dev_data->dma.cfg;
+	static DMA_HandleTypeDef hdma;
+	uint32_t ahb_clock_freq;
+	uint32_t prescaler = 0;
+	int ret;
+
+	/* Signals configuration */
+	ret = stm32_dt_pinctrl_configure(dev_cfg->pinctrl_list,
+					 dev_cfg->pinctrl_list_size,
+					 (uint32_t)dev_cfg->regs);
+	if (ret < 0) {
+		LOG_ERR("QSPI pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	/* Clock configuration */
+	__ASSERT_NO_MSG(device_get_binding(STM32_CLOCK_CONTROL_NAME));
+
+	if (clock_control_on(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			     (clock_control_subsys_t) &dev_cfg->pclken) != 0) {
+		LOG_DBG("Could not enable QSPI clock");
+		return -EIO;
+	}
+
+	if (clock_control_get_rate(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			(clock_control_subsys_t) &dev_cfg->pclken,
+			&ahb_clock_freq) < 0) {
+		LOG_DBG("Failed to get_AHB clock frequency");
+		return -EIO;
+	}
+
+	for (; prescaler <= STM32_QSPI_CLOCK_PRESCALER_MAX; prescaler++) {
+		uint32_t clk = ahb_clock_freq / (prescaler + 1);
+
+		if (clk <= dev_cfg->max_frequency) {
+			break;
+		}
+	}
+	__ASSERT_NO_MSG(prescaler <= STM32_QSPI_CLOCK_PRESCALER_MAX);
+
+	/*
+	 * DMA configuration
+	 * Due to use of QSPI HAL API in current driver,
+	 * both HAL and Zephyr DMA drivers should be configured.
+	 * The required configuration for Zephyr DMA driver should only provide
+	 * the minimum information to inform the DMA slot will be in used and
+	 * how to route callbacks.
+	 */
+
+	if (dev_data->dma.name != NULL) {
+		dev_data->dma.dev = device_get_binding(dev_data->dma.name);
+		if (!dev_data->dma.dev) {
+			LOG_ERR("%s device not found", dev_data->dma.name);
+			return -ENODEV;
+		}
+	}
+
+	/* Proceed to the minimum Zephyr DMA driver init */
+	dma_cfg.user_data = &hdma;
+	/* HACK: This field is used to inform driver that it is overridden */
+	dma_cfg.linked_channel = 0x7F;
+	ret = dma_config(dev_data->dma.dev, dev_data->dma.channel, &dma_cfg);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* Proceed to the HAL DMA driver init */
+	if (dma_cfg.source_data_size != dma_cfg.dest_data_size) {
+		LOG_ERR("Source and destination data sizes not aligned");
+		return -EINVAL;
+	}
+
+	int index = find_lsb_set(dma_cfg.source_data_size) - 1;
+
+	/* TODO: This init is limited to G0/G4/L0/L4/L5/WB */
+	hdma.Init.PeriphDataAlignment = table_p_size[index];
+	hdma.Init.MemDataAlignment = table_m_size[index];
+	hdma.Init.Request = dma_cfg.dma_slot;
+	hdma.Init.PeriphInc = DMA_PINC_DISABLE;
+	hdma.Init.MemInc = DMA_MINC_ENABLE;
+	hdma.Init.Mode = DMA_NORMAL;
+	hdma.Init.Priority = dma_cfg.channel_priority;
+#ifdef CONFIG_DMA_STM32_V1
+	hdma.Instance = __LL_DMA_GET_CHANNEL_INSTANCE(dev_data->dma.reg,
+						      dev_data->dma.channel);
+#else
+	hdma.Instance = __LL_DMA_GET_CHANNEL_INSTANCE(dev_data->dma.reg,
+						      dev_data->dma.channel-1);
+#endif /* CONFIG_DMA_STM32_V1 */
+
+	/* Initialize DMA HAL */
+	__HAL_LINKDMA(&dev_data->hqspi, hdma, hdma);
+	HAL_DMA_Init(&hdma);
```

Comment:
As far as I understand, Zephyr DMA API is not used here. What is the reason? Other STM32 drivers such as SPI don't do that, I wonder if the way non-LL HAL is built forces this choice.

Reply:
That's indeed the point.
DMA calls are done inside QSPI HAL functions, it not possible to unplug. And there is no QSPI LL API.
Alternative is to go with a full CMSIS port of the driver, but we'll lose abstraction capability (and would require significant dev and re-validation effort I can't afford on my side).
Current solution is a bit frustrating, but I think this is the best available option.

Reply:
@erwango got it. I think it's better to make more STM32 hw capabilities available before optimizing, so I think it's ok for now 👍